src/bartiq/analysis/rewriters/utils.py

```diff
@@ -61,12 +85,47 @@ def from_string(cls, assumption_string: str) -> Self:
         return cls(*_unpack_assumption(assumption_string))
 
 
+@dataclass(frozen=True)
+class Substitution(Instruction):
+    """Substitute a symbol with an expression."""
```

Comment:
[minor] class attributes lack description. In particular `wild`, it's unclear to me what it represents.
Also, I personally wouldn't call it `symbol_or_expr`, but rather just `expr` – `symbol` is just a special case of a "single symbol expression", no?